### PR TITLE
Hmascii Import

### PR DESCRIPTION
**Pull request by LinusKnecht**
Related Issues
----------------
Adapted from #21.

Description
----------------

This PR allows importing HyperMesh models from HMASCII files. It supports only 3D meshes with either hexahedron or tetrahedron elements.
It extends ```gustaf.io``` by a   ```hmascii``` tool. 

Comes with an example.

### CODE REVIEW

File: gustaf/io/hmascii.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        number_of_nodes: int
+        subelement: str
+        MeshType: Mesh
+
+        Returns
+        --------
+        None
+        """
+        self.number_of_nodes = int(number_of_nodes)
+        self.MeshType = MeshType
```
Comment (clemens-fricke) [inline gustaf/io/hmascii.py:73]: could you rename the `MeshType` variable name to be more inline with other variable names?

File: gustaf/io/hmascii.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    element_type: str
+        The volume element type to be imported. If this is not specified, the
+        most common element type in the main component is chosen.
+
+    Returns
+    -----------
+    mesh: Mesh
+    """
+
+    if not fname.endswith(".hmascii"):
```
Comment (clemens-fricke) [inline gustaf/io/hmascii.py:204]: Here this only really works if the name element is a pathlib/etc object and not with strings. Could you either make a type check or convert it if it is not of this type? 
pathlib does not care if a pathlib object is used as the init value meaning that 
```
fname = pathlib.Path(fname)
```

should work if fname is a pathlib.Path object and a string.

**Review by clemens-fricke — COMMENTED** [review]
I am not sure, but I think we previously did not really have mesh example files this need to be a discussion if we want to change this. @j042  

Otherwise very nice thank you.

**Comment by j042** [comment]
this should be rebased instead

**Review by j042 — APPROVED** [review]
thank you!

**Comment by clemens-fricke** [comment]
Needs https://github.com/tataratat/samples/pull/4 to be merged first due to a change of directory name.

File: gustaf/io/hmascii.py [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+                f"unknown element types {ignored_element_types}. "
+                "They will be ignored."
+            )
+
+        # are there volume elements? append.
+        if element_type in hm_component.elements:
+            elements_in_component = hm_component.elements[element_type]
+
+            # append elements
+            hm_volume_elements_nonunique = np.concatenate(
```
Comment (j042) [inline gustaf/io/hmascii.py:259]: this is quite expensive to have in a loop. can you just append to a list and concatenate once at the end?

File: gustaf/io/hmascii.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+    # create minimal vertex array
+    hm_node_indices = np.unique(volumes)
+    number_of_vertices = len(hm_node_indices)
+    if number_of_vertices == len(hm_model.node_coordinates):
+        vertices = np.array(hm_model.node_coordinates, dtype=np.float64)
+        node_perm = hm_model.node_ids
+    else:
+        # filter nodes
+        vertices = np.full((number_of_vertices, 3), 0.0)
```
Comment (j042) [inline gustaf/io/hmascii.py:295]:
```suggestion
        vertices = np.zeros((number_of_vertices, 3))
```

File: gustaf/io/hmascii.py [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+        node_perm = dict()
+        for vertex_id, hm_node_id in enumerate(hm_node_indices):
+            vertices[vertex_id, :] = hm_model.node_coordinates[
+                hm_model.node_ids[hm_node_id]
+            ]
+            node_perm[hm_node_id] = vertex_id
+
+    # finalize volumes array
```
Comment (j042) [inline gustaf/io/hmascii.py:303]: is this only for 3d?

**Pull request closed without merging**